File: bantora-common/bantora-common-persistence/src/main/java/com/t3ratech/bantora/persistence/entity/Poll.java
/**
 * Created by Cascade AI
 * Author       : Tsungai Kaviya
 * Copyright    : TeraTech Solutions (Pvt) Ltd
 * Date/Time    : 2025-11-28
 * Email        : tkaviya@t3ratech.co.zw
 */

package com.t3ratech.bantora.persistence.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "polls")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Poll {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "creator_phone", nullable = false, length = 20)
    private String creatorPhone;

    @Column(name = "idea_id")
    private UUID ideaId;

    @Enumerated(EnumType.STRING)
    @Column(name = "scope", nullable = false, length = 20)
    private PollScope scope;

    @Column(name = "region", length = 50)
    private String region;

    @Column(name = "country_code", length = 2)
    private String countryCode;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    @Builder.Default
    private PollStatus status = PollStatus.PENDING;

    @Column(name = "start_time", nullable = false)
    private Instant startTime;

    @Column(name = "end_time", nullable = false)
    private Instant endTime;

    @Column(name = "total_votes", nullable = false)
    @Builder.Default
    private Long totalVotes = 0L;

    @Column(name = "created_at", nullable = false, updatable = false)
    @Builder.Default
    private Instant createdAt = Instant.now();

    @Column(name = "updated_at", nullable = false)
    @Builder.Default
    private Instant updatedAt = Instant.now();

    @OneToMany(mappedBy = "poll", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<PollOption> options = new ArrayList<>();

    @PreUpdate
    public void preUpdate() {
        updatedAt = Instant.now();
    }
}

File: bantora-web/bantora_app/lib/models/poll.dart
class Poll {
  final String id;
  final String title;
  final String description;
  final String creatorPhone;
  final String? category;
  final String scope;
  final String? region;
  final String? countryCode;
  final String status;
  final DateTime createdAt;
  final DateTime? startTime;
  final DateTime? endTime;
  final List<PollOption> options;

  Poll({
    required this.id,
    required this.title,
    required this.description,
    required this.creatorPhone,
    this.category,
    required this.scope,
    this.region,
    this.countryCode,
    required this.status,
    required this.createdAt,
    this.startTime,
    this.endTime,
    required this.options,
  });

  int get totalVotes => options.fold(0, (sum, item) => sum + item.votesCount);

  factory Poll.fromJson(Map<String, dynamic> json) {
    return Poll(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      creatorPhone: json['creatorPhone'],
      category: json['category'],
      scope: json['scope'],
      region: json['region'],
      countryCode: json['countryCode'],
      status: json['status'],
      createdAt: DateTime.parse(json['createdAt']),
      startTime: json['startTime'] != null ? DateTime.parse(json['startTime']) : null,
      endTime: json['endTime'] != null ? DateTime.parse(json['endTime']) : null,
      options: (json['options'] as List)
          .map((option) => PollOption.fromJson(option))
          .toList(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'creatorPhone': creatorPhone,
      'scope': scope,
      'region': region,
      'countryCode': countryCode,
      'status': status,
      'createdAt': createdAt.toIso8601String(),
      'startTime': startTime?.toIso8601String(),
      'endTime': endTime?.toIso8601String(),
      'options': options.map((option) => option.toJson()).toList(),
    };
  }
}

class PollOption {
  final String id;
  final String pollId;
  final String optionText;
  final int optionOrder;
  final int votesCount;

  PollOption({
    required this.id,
    required this.pollId,
    required this.optionText,
    required this.optionOrder,
    required this.votesCount,
  });

  factory PollOption.fromJson(Map<String, dynamic> json) {
    return PollOption(
      id: json['id'],
      pollId: json['pollId'],
      optionText: json['optionText'],
      optionOrder: json['optionOrder'],
      votesCount: json['votesCount'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'pollId': pollId,
      'optionText': optionText,
      'optionOrder': optionOrder,
      'votesCount': votesCount,
    };
  }
}
